core/src/value/deserialize.rs

```diff
@@ -291,9 +301,146 @@ pub fn parse_datetime(buf: &[u8], dt_utc_offset: FixedOffset) -> Result<DateTime
+ *  Returns a tuple of two Option\<NaiveDate\>
+ *  None means no upper or lower range is present
+ */
+pub fn parse_date_range(buf: &[u8]) -> Result<(Option<NaiveDate>, Option<NaiveDate>)> {
+    // minimum length of one valid Date (YYYY) and one '-' separator
+    if buf.len() < 5 {
+        return UnexpectedEndOfElement.fail();
+    }
+
+    if let Some(separator) = buf.iter().position(|e| *e == b'-') {
+        let (start, end) = buf.split_at(separator);
+        let end = &end[1..];
+        match separator {
+            0 => Ok((None, Some(parse_date(end)?.0))),
+            i if i == buf.len() - 1 => Ok((Some(parse_date(start)?.0), None)),
+            _ => {
+                let (s, e) = (parse_date(start)?.0, parse_date(end)?.0);
+                check_range!(s, e)
+            }
+        }
+    } else {
+        NoRangeSeparator.fail()
+    }
+}
+
+/**
+ *  Looks for a range separator '-'.
+ *  Returns a tuple of two Option\<NaiveTime\>
+ *  None means no upper or lower range is present
+ */
+pub fn parse_time_range(buf: &[u8]) -> Result<(Option<NaiveTime>, Option<NaiveTime>)> {
+    // minimum length of one valid Time (HH) and one '-' separator
+    if buf.len() < 3 {
+        return UnexpectedEndOfElement.fail();
+    }
+
+    if let Some(separator) = buf.iter().position(|e| *e == b'-') {
+        let (start, end) = buf.split_at(separator);
+        let end = &end[1..];
+        match separator {
+            0 => Ok((None, Some(parse_time(end)?.0))),
+            i if i == buf.len() - 1 => Ok((Some(parse_time(start)?.0), None)),
+            _ => {
+                let (s, e) = (parse_time(start)?.0, parse_time(end)?.0);
+                check_range!(s, e)
+            }
+        }
+    } else {
+        NoRangeSeparator.fail()
+    }
+}
+
+/**
+ *  Looks for a range separator '-'.
+ *  Returns a tuple of two Option\<DateTime\>
+ *  None means no upper or lower range is present
+ */
+pub fn parse_datetime_range(
+    buf: &[u8],
+    dt_utc_offset: FixedOffset,
+) -> Result<(Option<DateTime<FixedOffset>>, Option<DateTime<FixedOffset>>)> {
+    // minimum length of one valid DateTime (YYYY) and one '-' separator
+    if buf.len() < 5 {
+        return UnexpectedEndOfElement.fail();
+    }
+
+    if let Some(separator) = buf
+        .iter()
+        .enumerate()
+        .find(|(i, c)| {
+            match **c == b'-' {
+                true => {
+                    match i {
+                        /* empty separator in the beginning */
+                        0 => true,
+                        /* empty separator at the end */
+                        x if *x == buf.len() - 1 => true,
+                        x if *x < buf.len() - 6 => {
+                            match buf[x + 5] {
+                                // separator present in 5 bytes, so assume this is an offset sign
+                                b'-' => false,
+                                _ => true,
+                            }
```

Comment:
This whole piece could turn into a single comparison, right?
```suggestion
                            // separator present in 5 bytes, so assume this is an offset sign
                            buf[x + 5] != b'-'
```